files/archive.py

```diff
@@ -0,0 +1,385 @@
+                module.fail_json(dest=dest, msg='Error deleting some source files: ' + str(e), files=errors)
+
+        # Rudimentary check: If size changed then file changed. Not perfect, but easy.
+        if os.path.getsize(dest) != size:
+            changed = True
+
+        if len(successes) and state != 'incomplete':
+            state = 'archive'
+
+    # Simple, single-file compression
+    else:
+        path = expanded_paths[0]
+
+        # No source or compressed file
+        if not (os.path.exists(path) or os.path.lexists(dest)):
+            state = 'absent'
+
+        # if it already exists and the source file isn't there, consider this done
+        elif not os.path.lexists(path) and os.path.lexists(dest):
+            state = 'compress'
+
+        else:
+            if module.check_mode:
+                if not os.path.exists(dest):
+                    changed = True
+            else:
+                size = 0
+                f_in = f_out = arcfile = None
+
+                if os.path.lexists(dest):
+                    size = os.path.getsize(dest)
+
+                try:
+                    if compression == 'zip':
+                        arcfile = zipfile.ZipFile(dest, 'w', zipfile.ZIP_DEFLATED)
+                        arcfile.write(path, path[len(arcroot):])
+                        arcfile.close()
+                        state = 'archive' # because all zip files are archives
+
+                    else:
+                        f_in = open(path, 'rb')
+
+                        if compression == 'gz':
+                            f_out = gzip.open(dest, 'wb')
+                        elif compression == 'bz2':
+                            f_out = bz2.BZ2File(dest, 'wb')
+                        else:
+                            raise OSError("Invalid compression")
+
+                        shutil.copyfileobj(f_in, f_out)
+
+                    successes.append(path)
+
+                except OSError:
+                    e = get_exception()
+
+                    module.fail_json(path=path, dest=dest, msg='Unable to write to compressed file: %s' % str(e))
+
+                if arcfile:
+                    arcfile.close()
+                if f_in:
+                    f_in.close()
+                if f_out:
+                    f_out.close()
+
+                # Rudimentary check: If size changed then file changed. Not perfect, but easy.
+                if os.path.getsize(dest) != size:
+                    changed = True
+
+            state = 'compress'
+
+        if remove:
+            try:
+                os.remove(path)
+
+            except OSError:
+                e = get_exception()
+                module.fail_json(path=path, msg='Unable to remove source file: %s' % str(e))
+
+    module.exit_json(archived=successes, dest=dest, changed=changed, state=state, arcroot=arcroot, missing=missing, expanded_paths=expanded_paths)
+
+# import module snippets
+from ansible.module_utils.basic import *
```

Comment:
Style note: This import can go near the top with the other imports now.  It no longer needs to be a wildcard import -- you can import exactly what you need (for instance, from ansible.moudle_utils.basic import AnsibleModule)